sdk/go/docker/docker.go

```diff
@@ -0,0 +1,558 @@
+// Copyright 2016-2020, Pulumi Corporation.
+
+package docker
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math/rand"
+	"os/exec"
+	"strings"
+
+	"github.com/Masterminds/semver"
+
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi/pkg/util/logging"
+	"github.com/pulumi/pulumi/sdk/go/pulumi"
+)
+
+func buildAndPushImageAsync(ctx *pulumi.Context, imageName string, build DockerBuild, repositoryURL string,
+	logResource pulumi.Resource, skipPush bool, registry ImageRegistry) (string, error) {
+
+	// Give an initial message indicating what we're about to do.  That way, if anything
+	// takes a while, the user has an idea about what's going on.
+	logging.Infof("Starting docker build and push...")
+
+	res, err := buildAndPushImageWorkerAsync(ctx, imageName, build, repositoryURL, logResource, skipPush, &registry)
```

Comment:
Why does `registry` turn from `registry ImageRegistry` to `registry *ImageRegistry` in this call?